CONTRIBUTING.md

```diff
@@ -51,6 +51,134 @@ If you're *looking* for security bugs, this crate is set up for
 PRs which cause test failures or a significant coverage decrease
 are unlikely to be accepted.
 
+## Style guide
+
+### Ordering
+
+#### Top-down ordering within modules
+
+Within a module, we prefer to order items top-down. This means that items within
+a module will depend on items defined below them, but not (usually) above them.
+The idea here is that the public API, with more internal dependencies, will be
+read (and changed) more often, and putting it closer to the top of the module
+makes it more accessible.
+
+This can be surprising to many engineers who are used to the bottom-up ordering
+used in languages like Python, where items can have a run-time dependency on
+other items defined in the same module.
+
+Usually `const` values will thus go on the bottom of the module (least complex,
+usually no dependencies of their own), although in larger modules it can make
+sense to place a `const` directly below the user (especially if there is a
+single user, or just a few co-located users).
+
+The `#[cfg(test)] mod tests {}` module goes on the very bottom, if present.
+
+#### Ordering for a given type
+
+For a given type, we prefer to order items as follows:
+
+1. The type definition (`struct` or `enum`)
+2. The inherent `impl` block (that is, not a trait implementation)
+3. `impl` blocks for traits, from most specific to least specific.
+   The least specific would be something like a `Debug` or `Clone` impl.
+
+#### Ordering associated functions within an inherent `impl` block
+
+Here's a guide to how we like to order associated functions:
+
+0. Associated functions (that is, `fn foo() {}` instead of `fn foo(&self) {}`)
+1. Constructors, starting with the constructor that takes the least arguments
+2. Public API that takes a `&mut self`
+3. Public API that takes a `&self`
+4. Private API that takes a `&mut self`
+5. Private API that takes a `&self`
+6. `const` values
+
+Note that we usually also practice top-down ordering here; where these are in
+conflict, make a choice that you think makes sense. For getters and setters, the
+order should typically mirror the order of the fields in the type definition.
+
+### Functions
+
+#### Consider avoiding short single-use functions
+
+We believe (properly commented) large functions can be more readable than
+splitting it up into many small single-use functions. Single-caller functions
+can be useful, but consider whether the code would be easier to follow if the
+function was inlined (especially for short functions).
```

Comment:
Hm, I don't feel fully on board with this still. Especially, I've been considering writing a PR which refactors `src/msgs/deframer.rs` into a style where the comments that introduce sections of code are in individual singler-caller functions that can be understood incrementally, as opposed to a single 200-line one which has to be understood as a whole.

As an illustration as to why I think single-use functions are better than comments: there's a comment there which talks about "place the message onto the frames output queue", but there is no such queue now. If there was a single-use function like `fn deframe_one_message_if_present(output: &mut VecDequeue<Message>)` (or whatever) the compiler would be involved in ensuring the structural documentation (the names, arguments, returns of the smaller functions) made sense.

Reply:
I agree the deframer is pretty gnarly, although I think it's not very representative of the larger point (with its nested loops). This part of the style guide tries to thread the needle of making recommendations for the common case without imposing strict rules -- I know there have certainly been other areas where I did go for single-use functions to make the high-level flow clearer. Do you want to try to reword this into something that you'd feel more comfortable with?

For your specific example, it sounds like that comment might just be outdated, which might be a separate issue -- but I'm also curious how you would structure the code. It can also sometimes be tricky to get the borrow checker to handle splitting up code like this.

Reply:
As a discussion point, consider something more like `server::tls13::client_hello::handle_client_hello()`. Do you also feel that should get split up into smaller functions?

Reply:
> Do you also feel that should get split up into smaller functions?

Yes, I think so. A prime candidate there would be the block around where `resumedata` needs to be `mut`. If all that was in a function that fully decides on `resumedata`/`chosen_psk_index`, I think it would be clearer.

I don't think I'd extract more than one or two functions from there though -- I think it's beneficial for that function if it contains all the possible transitions to subsequent states in the function itself. But quite a lot of it is of the form `if `client_hello.something().has_some_required_property() { return Err(cx.common.send_fatal_alert(...)); }`.

Reply:
That's fair, I think both the `chosen_share == None` and `psk` handling are kinda clear candidates here. So do you think we should just strike this language, or will it still be useful if weakened form?

Reply:
>  or will it still be useful if weakened form?

I don't have a strong opinion but I think there's still utility in a weakened form.

Reply:
Here's a somewhat softer take, what do you think?

"While single-use functions can make sense if the algorithm is sufficiently complex that it warrants an explicit name and interface, using many short single-use functions can make the code harder to follow, due to having to jump around in order to gain an understanding of what's going on. When writing a single-use function, consider whether it needs the dedicated interface, or if it could be inlined into its caller instead."